src/libraries/System.IO.Packaging/tests/PartPieceTests.cs

```diff
@@ -0,0 +1,606 @@
+
+            ms.Write(_partPieceSampleZipPackage);
+            var zipArchive = new ZipArchive(ms, ZipArchiveMode.Update);
+            var zipPackage = Package.Open(ms, FileMode.Open);
+
+            var atomicPart = zipPackage.GetPart(atomicPartEntryUri);
+            var writeStream = atomicPart.GetStream(FileMode.OpenOrCreate);
+
+            writeStream.Write([0, 0]);
+            writeStream.Seek(16, SeekOrigin.Begin);
+            writeStream.Write([0, 0]);
+
+            writeStream.Flush();
+
+            zipPackage.Flush();
+            (zipPackage as IDisposable).Dispose();
+
+            ms.Seek(0, SeekOrigin.Begin);
+
+            zipArchive = new ZipArchive(ms, ZipArchiveMode.Read);
+            zipPackage = Package.Open(ms, FileMode.Open);
+            atomicPart = zipPackage.GetPart(atomicPartEntryUri);
+            using var readStream = atomicPart.GetStream(FileMode.OpenOrCreate);
+            var buffer = new byte[24];
+
+            readStream.Read(buffer);
+
+            Assert.Equal(buffer, [0, 0, 2, 3, 4, 5, 6, 7, 8, 9, 10, 11, 12, 13, 14, 15, 0, 0, 18, 19, 20, 21, 22, 23]);
+        }
+
+        [Theory]
+        [InlineData("ReadablePartPieceEntry.bin", 30, "[0].piece,[1].piece,[2].last.piece", "[2].piece,[3].last.piece")]
+        [InlineData("ReadablePartPieceEntry.bin", 0, "[0].last.piece", "[0].piece,[1].piece,[2].piece,[3].last.piece")]
+        public void SetLengthRemovesPartPieces(string partName, int newLength, string remainingPieces, string removedPieces)
+        {
+            using var ms = new MemoryStream();
+
+            ms.Write(_partPieceSampleZipPackage);
+            var zipArchive = new ZipArchive(ms, ZipArchiveMode.Update);
+            var zipPackage = Package.Open(ms, FileMode.Open);
+
+            var partEntry = zipPackage.GetPart(new Uri("/" + partName, UriKind.Relative));
+            var writeStream = partEntry.GetStream(FileMode.OpenOrCreate);
+
+            // We perform a write to force the InterleavedPartStream to lazily initialise some of the underlying streams.
+            // Doing so forces ZipPackage to open the ZipArchiveEntry (which in turn would normally prevent it from being deleted.)
+            writeStream.Write([0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0]);
+            writeStream.SetLength(newLength);
+
+            writeStream.Flush();
+
+            zipPackage.Flush();
+            (zipPackage as IDisposable).Dispose();
+
+            ms.Seek(0, SeekOrigin.Begin);
+
+            zipArchive = new ZipArchive(ms, ZipArchiveMode.Read);
+
+            foreach (var pieceName in remainingPieces.Split(','))
+            {
+                Assert.NotNull(zipArchive.GetEntry(partName + "/" + pieceName));
+            }
+            foreach (var pieceName in removedPieces.Split(','))
+            {
+                Assert.Null(zipArchive.GetEntry(partName + "/" + pieceName));
+            }
+
+            zipArchive.Dispose();
+        }
+    }
+}
```

Comment:
Can you please add at least one test for an unseekable stream? We need to confirm this code works for, say, a network stream as an input.

You can continue using a MemoryStream as an input, but it needs to be wrapped with [a WrappedStream](https://github.com/dotnet/runtime/blob/cae3aec80d32316026d85228e12a94dd4887d57f/src/libraries/Common/tests/System/IO/WrappedStream.cs). There are many examples in System.Formats.Tar, like [this one](https://github.com/dotnet/runtime/blob/cae3aec80d32316026d85228e12a94dd4887d57f/src/libraries/System.Formats.Tar/tests/TarWriter/TarWriter.WriteEntry.Tests.cs#L544).

Reply:
Would you mind clarifying please? The part pieces are built on `ZipArchiveEntry` objects, and the streams returned by `ZipArchiveEntry` are always non-seekable on .NET Core (even if the stream the Package's `ZipArchive` was opened from is seekable.)

Reply:
I understand now, thanks. I'd expect different behaviour though, generated by `ZipArchive`.

| Package.Open `packageAccess` | Resultant ZipArchiveMode | Non-seekable stream behaviour |
| - | - | - |
| `FileAccess.Read` | `ZipArchiveMode.Read` | Stream copied to (seekable) MemoryStream |
| `FileAccess.Write` | `ZipArchiveMode.Create` | No impact - can only be used to create Packages |
| `FileAccess.ReadWrite` (default) | `ZipArchiveMode.Update` | ZipArchive constructor will throw |

I expect this'll throw within [the ZipArchive constructor](https://github.com/dotnet/runtime/blob/main/src/libraries/System.IO.Compression/src/System/IO/Compression/ZipArchive.cs#L120) before it even reaches the code under test in Package. I'll add tests which verify these behaviours.

Reply:
I see your point now, thanks - there's a behavioural difference between `ZipArchive` on .NET Core and Framework. We're now testing both behaviours, bringing us to three tests:
1. (.NET Core) Can create a package and add a part when the package is backed by an unseekable stream.
2. (.NET Framework) Cannot create a package when the package is backed by an unseekable stream.
3. Cannot open a package for modification when the package is backed by an unseekable stream.

Once the tests complete in CI, I'll mark this as resolved.